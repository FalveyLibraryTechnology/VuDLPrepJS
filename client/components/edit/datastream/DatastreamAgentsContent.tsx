import React, { useEffect, useLayoutEffect, useRef, useState } from "react";
import Button from "@mui/material/Button";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import Grid from "@mui/material/Grid";
import { useEditorContext } from "../../../context/EditorContext";
import { useGlobalContext } from "../../../context/GlobalContext";
import useDatastreamOperation from "../../../hooks/useDatastreamOperation";
import CircularProgress from "@mui/material/CircularProgress";
import Divider from "@mui/material/Divider";
import DatastreamAgentsAddContentRow from "./DatastreamAgentsAddContentRow";
import DatastreamAgentsModifyContentRow from "./DatastreamAgentsModifyContentRow";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";
import Tab from "@mui/material/Tab";
import Tabs from "@mui/material/Tabs";
import PidPicker from "../PidPicker";

const DatastreamAgentsContent = (): React.ReactElement => {
    const {
        state: { agentsCatalog, currentAgents, objectDetailsStorage },
        action: { loadObjectDetailsIntoStorage, setCurrentAgents },
    } = useEditorContext();
    const {
        action: { closeModal },
    } = useGlobalContext();
    const { uploadAgents, getAgents } = useDatastreamOperation();
    const {
        defaults: { role, type, name },
    } = agentsCatalog;
    const [addAgent, setAddAgent] = useState({
        role: "",
        type: "",
        name: "",
        notes: [],
    });
    const [hasChanges, setHasChanges] = useState(false);
    const [isLoading, setIsLoading] = useState(true);
    const [clonePid, setClonePid] = useState("");
    const EDIT_TAB = 0;
    const CLONE_TAB = 1;
    const [tab, setTab] = useState<number>(EDIT_TAB);
    const handleTabChange = (event: React.SyntheticEvent, newValue: number) => {
        setTab(newValue);
    };
    const clonePidLoaded = clonePid.length > 0 && Object.prototype.hasOwnProperty.call(objectDetailsStorage, clonePid);
    const loadClonePid = async (newClonePid: string) => {
        if (newClonePid.length == 0) {
            setClonePid("");
            return;
        }
        if (!Object.prototype.hasOwnProperty.call(objectDetailsStorage, newClonePid)) {
            let error = false;
            const errorCallback = () => {
                error = true;
            };
            await loadObjectDetailsIntoStorage(newClonePid, errorCallback);
            if (error) {
                alert(`Cannot load PID: ${newClonePid}`);
                return;
            }
        }
        setClonePid(newClonePid);
    };
    const canSave = currentAgents.every(({ role, type, name }) => {
        return role && type && name;
    });
    const canSaveAddAgent = addAgent.role && addAgent.type && addAgent.name;
    const saveCurrentAgents = async () => {
        setIsLoading(true);
        let agents = currentAgents;
        if (canSaveAddAgent) {
            agents = [...agents, addAgent];
            setAddAgent({
                role: "",
                type: "",
                name: "",
                notes: [],
            });
            setCurrentAgents(agents);
        }
        await uploadAgents(agents);
        setIsLoading(false);
    };
    const contentRef = useRef();

    useLayoutEffect(() => {
        if (contentRef.current) {
            contentRef.current.scrollTop = contentRef.current.scrollHeight;
        }
    }, [contentRef]);

    const callGetAgents = async (overridePid: string | null = null, force = false) => {
        const currentAgents = await getAgents(overridePid, force);
        if (!currentAgents.length) {
            currentAgents.push({ role, type, name, notes: [] });
            setHasChanges(true);
        }
        setCurrentAgents(currentAgents);
        setIsLoading(false);
    };
    useEffect(() => {
        callGetAgents();
    }, []);
    const doClone = async () => {
        const details = objectDetailsStorage[clonePid] ?? {};
        if ((details?.datastreams ?? []).includes("AGENTS")) {
            setIsLoading(true);
            await callGetAgents(clonePid, true);
        } else {
            alert(`${clonePid} does not contain an AGENTS datastream.`);
        }
        setTab(EDIT_TAB);
        setClonePid("");
    };
    return (
        <>
            <DialogContent sx={{ width: "100%", minHeight: "50vh" }} ref={contentRef}>
                {isLoading ? (
                    <Box display="flex" justifyContent="center" alignItems="center" minHeight="50vh">
                        <CircularProgress />
                    </Box>
                ) : (
                    <>
                        <Box sx={{ borderBottom: 1, borderColor: "divider", marginBottom: "1em" }}>
                            <Tabs value={tab} onChange={handleTabChange}>
                                <Tab label="Editor" />
                                <Tab label="Clone" />
                            </Tabs>
                        </Box>
                        <div style={{ display: tab == EDIT_TAB ? "block" : "none" }}>
                            <Grid container spacing={1} sx={{ padding: "20px 24px" }}>
                                <Grid container item xs={3}>
                                    Role
                                </Grid>
                                <Grid container item xs={3}>
                                    Type
                                </Grid>
                                <Grid container item xs={5}>
                                    Name
                                </Grid>
                                <Grid container item xs={1}>
                                    Actions
                                </Grid>
                            </Grid>
                            <Grid container spacing={1}>
                                {currentAgents.map((agent, index) => {
                                    return (
                                        <DatastreamAgentsModifyContentRow
                                            key={index}
                                            agent={agent}
                                            index={index}
                                            setHasChanges={setHasChanges}
                                        />
                                    );
                                })}
                                <DatastreamAgentsAddContentRow
                                    agent={addAgent}
                                    setAgent={setAddAgent}
                                    setHasChanges={setHasChanges}
                                />
                            </Grid>
                        </div>
                        <div style={{ display: tab == CLONE_TAB ? "block" : "none" }}>
                            <div>
                                <PidPicker selected={clonePid} setSelected={loadClonePid} />
                            </div>
                            {clonePidLoaded ? <button onClick={doClone}>Clone</button> : null}
                        </div>
                    </>
                )}
            </DialogContent>
            <Divider />
            <DialogActions sx={{ padding: "20px 24px" }}>
                <Grid container spacing={1} justifyContent="flex-end">
                    <Grid container item xs={6}>
                        {hasChanges && (
                            <Typography
                                variant="subtitle1"
                                sx={{
                                    color: canSave ? "text.secondary" : "error.main",
                                    display: "flex",
                                    flexDirection: "column",
                                    justifyContent: "center",
                                }}
                            >
                                {canSave
                                    ? "You possibly have unsaved changes. Please save them."
                                    : "Role, Type, or Name field from existing rows cannot be empty."}
                            </Typography>
                        )}
                    </Grid>
                    <Grid container item xs={6} justifyContent="flex-end">
                        {hasChanges && (
                            <Button
                                className="agentsSaveChangesButton"
                                disabled={!canSave}
                                onClick={async () => {
                                    await saveCurrentAgents();
                                    setHasChanges(false);
                                }}
                            >
                                Save Changes
                            </Button>
                        )}
                        {hasChanges && (
                            <Button
                                className="agentsSaveCloseButton"
                                disabled={!canSave}
                                onClick={async () => {
                                    await saveCurrentAgents();
                                    closeModal("datastream");
                                }}
                            >
                                Save And Close
                            </Button>
                        )}

                        <Button
                            className="agentsCancelButton"
                            onClick={async () => {
                                setCurrentAgents(await getAgents());
                                closeModal("datastream");
                            }}
                        >
                            Cancel
                        </Button>
                    </Grid>
                </Grid>
            </DialogActions>
        </>
    );
};

export default DatastreamAgentsContent;
