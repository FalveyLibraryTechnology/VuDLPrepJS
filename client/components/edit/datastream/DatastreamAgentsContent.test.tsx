import React from "react";
import { describe, beforeEach, afterEach, expect, it, jest } from "@jest/globals";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import DatastreamAgentsContent from "./DatastreamAgentsContent";

const mockDatastreamAgentsModifyContentRow = jest.fn();
jest.mock("./DatastreamAgentsModifyContentRow", () => (props) => {
    mockDatastreamAgentsModifyContentRow(props);
    return "DatastreamAgentsModifyContentRow";
});
const mockDatastreamAgentsAddContentRow = jest.fn();
jest.mock("./DatastreamAgentsAddContentRow", () => (props) => {
    mockDatastreamAgentsAddContentRow(props);
    return "DatastreamAgentsAddContentRow";
});
const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
const mockUseGlobalContext = jest.fn();
jest.mock("../../../context/GlobalContext", () => ({
    useGlobalContext: () => {
        return mockUseGlobalContext();
    },
}));
const mockUseDatastreamOperation = jest.fn();
jest.mock("../../../hooks/useDatastreamOperation", () => () => {
    return mockUseDatastreamOperation();
});

let tabChangeFunction: ((tab: number) => void) | null = null;
jest.mock("@mui/material/Tabs", () => (props) => {
    tabChangeFunction = props.onChange;
    return props.children;
});
jest.mock("@mui/material/Tab", () => (props) => `Tab: ${props.label}`);
jest.mock("@mui/material/Grid", () => (props) => props.children);
let pidPickerFunction: ((pid: string) => void) | null = null;
jest.mock("../PidPicker", () => (props) => {
    pidPickerFunction = props.setSelected;
    return "PidPicker: " + JSON.stringify(props);
});

describe("DatastreamAgentsContent", () => {
    let editorValues;
    let globalValues;
    let datastreamOperationValues;
    beforeEach(() => {
        editorValues = {
            state: {
                agentsCatalog: {
                    defaults: {
                        role: "test1",
                        type: "test2",
                        name: "test3",
                    },
                },
                currentAgents: [],
                objectDetailsStorage: {},
            },
            action: {
                setCurrentAgents: jest.fn(),
                loadObjectDetailsIntoStorage: jest.fn(),
            },
        };
        globalValues = {
            action: {
                closeModal: jest.fn(),
            },
        };
        datastreamOperationValues = {
            uploadAgents: jest.fn(),
            getAgents: jest.fn(),
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseGlobalContext.mockReturnValue(globalValues);
        mockUseDatastreamOperation.mockReturnValue(datastreamOperationValues);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    async function renderComponent() {
        await act(async () => {
            render(<DatastreamAgentsContent />);
        });
    }

    it("renders, and calls getAgents on render", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        let tree;
        await renderer.act(async () => {
            tree = renderer.create(<DatastreamAgentsContent />);
        });
        await waitFor(() => expect(datastreamOperationValues.getAgents).toHaveBeenCalled());
        expect(editorValues.action.setCurrentAgents).toHaveBeenCalled();
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("supports tab switching", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        let tree;
        await renderer.act(async () => {
            tree = renderer.create(<DatastreamAgentsContent />);
        });
        await waitFor(() => expect(datastreamOperationValues.getAgents).toHaveBeenCalled());
        await renderer.act(async () => {
            tabChangeFunction && tabChangeFunction(null, 1);
        });
        expect(editorValues.action.setCurrentAgents).toHaveBeenCalled();
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("saves current changes on save changes click", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);

        await renderComponent();
        await waitFor(() => expect(datastreamOperationValues.getAgents).toHaveBeenCalled());
        await userEvent.setup().click(screen.getByText("Save Changes"));

        expect(datastreamOperationValues.uploadAgents).toHaveBeenCalled();
        expect(globalValues.action.closeModal).not.toHaveBeenCalled();
    });

    it("saves current agents and closes the modal", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);

        await renderComponent();
        await waitFor(() => expect(datastreamOperationValues.getAgents).toHaveBeenCalled());
        await userEvent.setup().click(screen.getByText("Save And Close"));

        expect(datastreamOperationValues.uploadAgents).toHaveBeenCalled();
        expect(globalValues.action.closeModal).toHaveBeenCalledWith("datastream");
    });

    it("resets current agents on cancel", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);

        await renderComponent();
        await waitFor(() => expect(datastreamOperationValues.getAgents).toHaveBeenCalled());
        await userEvent.setup().click(screen.getByText("Cancel"));

        expect(datastreamOperationValues.uploadAgents).not.toHaveBeenCalled();
        expect(datastreamOperationValues.getAgents).toHaveBeenCalled();
        expect(editorValues.action.setCurrentAgents).toHaveBeenCalled();
        expect(globalValues.action.closeModal).toHaveBeenCalledWith("datastream");
    });

    it("ignores empty PIDs in clone input", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        await renderComponent();
        expect(pidPickerFunction).not.toBeNull();
        await act(async () => {
            pidPickerFunction && pidPickerFunction("");
        });
        expect(editorValues.action.loadObjectDetailsIntoStorage).not.toHaveBeenCalled();
    });

    it("handles errors that occur during PID cloning", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        await renderComponent();
        const alertSpy = jest.spyOn(window, "alert").mockImplementation(jest.fn());
        editorValues.action.loadObjectDetailsIntoStorage.mockImplementation(
            (pid: string, errorCallback: () => void) => {
                expect(pid).toEqual("foo:123");
                errorCallback();
            },
        );
        expect(pidPickerFunction).not.toBeNull();
        await act(async () => {
            pidPickerFunction && pidPickerFunction("foo:123");
        });
        expect(alertSpy).toHaveBeenCalledWith("Cannot load PID: foo:123");
    });

    it("supports loading PID data into storage after selection for cloning", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        await renderComponent();
        expect(pidPickerFunction).not.toBeNull();
        await act(async () => {
            pidPickerFunction && pidPickerFunction("foo:123");
        });
        expect(editorValues.action.loadObjectDetailsIntoStorage).toHaveBeenCalledWith("foo:123", expect.anything());
    });

    it("validates datastreams before cloning metadata from another PID", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        editorValues.state.objectDetailsStorage = {
            "foo:123": {
                datastreams: [],
            },
        };
        const alertSpy = jest.spyOn(window, "alert").mockImplementation(jest.fn());
        await renderComponent();
        expect(pidPickerFunction).not.toBeNull();
        await act(async () => {
            pidPickerFunction && pidPickerFunction("foo:123");
        });
        await userEvent.setup().click(screen.getByText("Clone"));
        expect(alertSpy).toHaveBeenCalledWith("foo:123 does not contain an AGENTS datastream.");
    });

    it("can clone metadata from another PID", async () => {
        datastreamOperationValues.getAgents.mockResolvedValue([]);
        editorValues.state.objectDetailsStorage = {
            "foo:123": {
                datastreams: ["AGENTS"],
            },
        };
        await renderComponent();
        expect(pidPickerFunction).not.toBeNull();
        await act(async () => {
            pidPickerFunction && pidPickerFunction("foo:123");
        });
        await userEvent.setup().click(screen.getByText("Clone"));
        expect(datastreamOperationValues.getAgents).toHaveBeenCalledWith("foo:123", true);
    });
});
