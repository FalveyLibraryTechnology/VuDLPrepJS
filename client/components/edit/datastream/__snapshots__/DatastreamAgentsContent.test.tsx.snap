// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatastreamAgentsContent renders, and calls getAgents on render 1`] = `
[
  <div
    className="MuiDialogContent-root css-pvro84-MuiDialogContent-root"
  >
    <div
      className="MuiBox-root css-1hjmj0z"
    >
      Tab: Editor
      Tab: Clone
    </div>
    <div
      style={
        {
          "display": "block",
        }
      }
    >
      Role
      Type
      Name
      Actions
      DatastreamAgentsAddContentRow
    </div>
    <div
      style={
        {
          "display": "none",
        }
      }
    >
      <div>
        PidPicker: {"selected":""}
      </div>
    </div>
  </div>,
  <hr
    className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
  />,
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing css-1wawfzm-MuiDialogActions-root"
  >
    <h6
      className="MuiTypography-root MuiTypography-subtitle1 css-1tvjuyl-MuiTypography-root"
    >
      You possibly have unsaved changes. Please save them.
    </h6>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary agentsSaveChangesButton css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Save Changes
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary agentsSaveCloseButton css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Save And Close
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary agentsCancelButton css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Cancel
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
]
`;
